File: mac/Sources/LlmIdeMac/Services/MemoryService.swift
// High-level memory service: typed read/write + validation over the Phase 1
// MemoryStorage layer. This is the service-tier surface other Mac-app code
// (and later service tasks) consume; it delegates all file I/O to
// MemoryStorage and adds graceful degradation (never crashes on read —
// returns empty data / logs) and fact validation.
//
// Mirrors the TS service layer (Task 4 of the storage/service split). The
// bugs/qa fields of MemoryData are forward-looking placeholders — fault/QA
// ingestion lands in a later phase — so they are always empty for now.
//
// Deviations from the original task template (forced by the real Phase 1
// types so the code compiles with no breaking changes):
//   - ChatMemoryFact.timestamp is `Int` (ms), not Double.
//   - QAEntry already exists (CodeGraph) and is not Codable, so MemoryData is
//     Sendable-only (Codable dropped); it is implicitly Sendable because all
//     stored properties are Sendable.
//   - BugReport is undefined in the module today, so a minimal Codable &
//     Sendable placeholder is declared here for the `bugs` slot.

import Foundation

/// Memory service protocol for high-level memory operations.
///
/// Conforming types must be `Sendable` (the bundled impl is an actor) so a
/// single shared instance is safe to reuse across the app.
protocol MemoryService: Sendable {
    /// Read the full memory snapshot for a repo (facts today; bugs/qa in a
    /// later phase). Degrades gracefully — returns empty data on read failure
    /// rather than throwing.
    func readMemory(repoRoot: URL) async throws -> MemoryData

    /// Read the chat-memory facts for a repo. Returns `[]` when absent or
    /// unreadable.
    func readChatMemory(repoRoot: URL) async throws -> [ChatMemoryFact]

    /// Persist chat-memory facts, overwriting any existing file atomically.
    func writeChatMemory(repoRoot: URL, facts: [ChatMemoryFact]) async throws

    /// Validate a fact without writing it. Checks text length (<=280 chars,
    /// matching the TS contract) and that every referenced file exists under
    /// `repoRoot`. Returns the accumulated errors; `valid` is true iff there
    /// are none.
    func validateFact(repoRoot: URL, fact: ChatMemoryFact) async throws -> ValidationResult

    /// Overwrite `repo.md` with `content` (atomic write via MemoryStorage).
    func updateRepoMD(repoRoot: URL, content: String) async throws
}

/// Memory data container returned by `readMemory`.
///
/// `bugs` and `qa` are placeholders for later phases (fault + Q&A ingestion)
/// and are empty today. `Sendable` so it can cross the actor boundary.
struct MemoryData: Sendable {
    let facts: [ChatMemoryFact]
    let bugs: [BugReport]
    let qa: [QAEntry]

    init(facts: [ChatMemoryFact] = [], bugs: [BugReport] = [], qa: [QAEntry] = []) {
        self.facts = facts
        self.bugs = bugs
        self.qa = qa
    }
}

/// Validation outcome for `validateFact`.
struct ValidationResult: Codable, Sendable {
    let valid: Bool
    let errors: [String]
}

/// Placeholder for a bug/fault entry. The faults pipeline (a later phase)
/// will populate this; for now it exists so `MemoryData` has a concrete,
/// `Codable` & `Sendable` slot for `bugs`.
struct BugReport: Codable, Sendable {}

/// Memory service implementation. An actor for thread safety; all file I/O
/// is delegated to the injected `MemoryStorage` (Phase 1 storage layer).
final actor MemoryServiceImpl: MemoryService {
    private let storage: MemoryStorage

    init(storage: MemoryStorage = MemoryStorage()) {
        self.storage = storage
    }

    func readMemory(repoRoot: URL) async throws -> MemoryData {
        do {
            // Read repo.md too so this stays the one "load all memory" entry
            // point. The content isn't surfaced in MemoryData yet (no field),
            // but performing the read keeps the call shape ready for that
            // field and mirrors the service-tier contract.
            _ = try? await storage.readMemoryFile(repoRoot: repoRoot, filename: "repo.md")
            let chatMemory = try await storage.readChatMemory(repoRoot: repoRoot)
            return MemoryData(facts: chatMemory, bugs: [], qa: [])
        } catch {
            // Graceful degradation — never crash; return empty data.
            return MemoryData(facts: [], bugs: [], qa: [])
        }
    }

    func readChatMemory(repoRoot: URL) async throws -> [ChatMemoryFact] {
        do {
            return try await storage.readChatMemory(repoRoot: repoRoot)
        } catch {
            // Graceful degradation — log and return empty rather than throw.
            print("Chat memory read failed: \(error)")
            return []
        }
    }

    func writeChatMemory(repoRoot: URL, facts: [ChatMemoryFact]) async throws {
        try await storage.writeChatMemory(repoRoot: repoRoot, facts: facts)
    }

    func validateFact(repoRoot: URL, fact: ChatMemoryFact) async throws -> ValidationResult {
        var errors: [String] = []

        // Text length (TS contract: 280 chars max).
        if fact.text.count > 280 {
            errors.append("Fact text exceeds 280 characters")
        }

        // Referenced files must exist under the repo root.
        if let files = fact.metadata?.files {
            for fileRef in files {
                let fullPath = repoRoot.appendingPathComponent(fileRef)
                if FileManager.default.fileExists(atPath: fullPath.path) == false {
                    errors.append("Referenced file does not exist: \(fileRef)")
                }
            }
        }

        return ValidationResult(valid: errors.isEmpty, errors: errors)
    }

    func updateRepoMD(repoRoot: URL, content: String) async throws {
        try await storage.writeMemoryFile(repoRoot: repoRoot, filename: "repo.md", content: content)
    }
}

File: extension/graphkit/services/memory-service.ts
// MemoryService: high-level memory operations with validation.
//
// This is the Phase 2 service layer. It uses the Phase 1 storage functions
// (storage/memory-storage.ts) internally for all file I/O while presenting a
// stable, validated API to callers. Reads degrade gracefully (never throw);
// writes surface failures so callers can react.
//
// NOTE on type parity: the service returns the Phase 1 `ValidationResult` and
// `ValidationReport` shapes verbatim (reason/details and valid/invalid/errors
// respectively). An earlier draft of this service planned richer
// `{ valid, errors: string[] }` / `{ total, valid, invalid, details }` shapes,
// but those do not match the shipped Phase 1 types; the implementation below
// honors the real types. Multi-message validation detail is carried in
// `ValidationResult.details` (joined by '; ').

import { promises as fs } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
// Imports use `.ts` specifiers (not `.js`) so the module loads under
// `node --test --experimental-strip-types` (Node >= 20). The repo tsconfig sets
// `allowImportingTsExtensions: true`, so the bundler / `tsc` accept these too.
import { writeMemoryFile, readChatMemory, writeChatMemory } from '../storage/memory-storage.ts';
import type {
  MemoryData,
  ChatMemoryFact,
  ValidationResult,
  ValidationReport
} from '../types/memory.ts';

/**
 * MemoryService provides high-level memory operations with validation.
 *
 * Initially built on top of the Phase 1 storage layer; repo.md/chat-memory.md
 * rendering still flows through the storage layer (the legacy memory.mjs
 * `renderRepoMemory` symbol referenced in some drafts does not exist in the
 * shipped module, so no delegation import is taken here).
 */
export class MemoryService {
  /**
   * Read all memory data for a repo.
   *
   * Reads degrade gracefully: if the memory directory or files are absent the
   * service returns an empty `MemoryData` rather than throwing.
   *
   * (repo.md is user-curated content with no field on `MemoryData`; bug/QA
   * reading is deferred to a later phase, so those arrays are returned empty.)
   */
  async readMemory(repoRoot: URL): Promise<MemoryData> {
    try {
      const chatMemory = await readChatMemory(repoRoot);
      return {
        facts: chatMemory,
        bugs: [], // TODO: implement bug reading in a later phase
        qa: []    // TODO: implement QA reading in a later phase
      };
    } catch (err) {
      // Graceful degradation — never crash the agent.
      console.error('Memory read failed:', err);
      return { facts: [], bugs: [], qa: [] };
    }
  }

  /**
   * Read chat memory facts only.
   */
  async readChatMemory(repoRoot: URL): Promise<ChatMemoryFact[]> {
    try {
      return await readChatMemory(repoRoot);
    } catch (err) {
      console.error('Chat memory read failed:', err);
      return [];
    }
  }

  /**
   * Write chat memory facts.
   *
   * Unlike reads, a failed write is re-thrown after logging so callers can
   * surface the failure (silently dropping a write would corrupt memory state).
   */
  async writeChatMemory(repoRoot: URL, facts: ChatMemoryFact[]): Promise<void> {
    try {
      await writeChatMemory(repoRoot, facts);
    } catch (err) {
      console.error('Chat memory write failed:', err);
      throw err;
    }
  }

  /**
   * Validate a single fact.
   *
   * Checks text length and that any referenced files exist beneath the repo
   * root. Problems are accumulated and returned in `details` (joined by '; ');
   * `reason` is set to the most specific applicable code (currently
   * 'file_not_found' when a referenced file is missing).
   */
  async validateFact(repoRoot: URL, fact: ChatMemoryFact): Promise<ValidationResult> {
    const problems: string[] = [];
    let reason: ValidationResult['reason'];

    // Check text length.
    if (fact.text.length > 280) {
      problems.push('Fact text exceeds 280 characters');
    }

    // Check file references exist beneath the repo root.
    const fileRefs = fact.metadata?.files;
    if (fileRefs && fileRefs.length > 0) {
      const repoPath = fileURLToPath(repoRoot);
      for (const fileRef of fileRefs) {
        const fullPath = path.join(repoPath, fileRef);
        try {
          await fs.access(fullPath);
        } catch {
          problems.push(`Referenced file does not exist: ${fileRef}`);
          reason = 'file_not_found';
        }
      }
    }

    return {
      valid: problems.length === 0,
      reason: problems.length === 0 ? undefined : reason,
      details: problems.length === 0 ? undefined : problems.join('; ')
    };
  }

  /**
   * Validate all facts currently stored in chat memory.
   */
  async validateAllFacts(repoRoot: URL): Promise<ValidationReport> {
    const facts = await this.readChatMemory(repoRoot);
    const results = await Promise.all(
      facts.map((fact) => this.validateFact(repoRoot, fact))
    );

    const errors: ValidationReport['errors'] = [];
    results.forEach((result, i) => {
      if (!result.valid) {
        errors.push({
          fact: facts[i],
          reason: result.details ?? result.reason ?? 'invalid'
        });
      }
    });

    return {
      valid: results.filter((r) => r.valid).length,
      invalid: results.filter((r) => !r.valid).length,
      errors
    };
  }

  /**
   * Update repo.md user-curated facts.
   *
   * Failures are re-thrown after logging (see writeChatMemory).
   */
  async updateRepoMD(repoRoot: URL, content: string): Promise<void> {
    try {
      await writeMemoryFile(repoRoot, 'repo.md', content);
    } catch (err) {
      console.error('Repo.md update failed:', err);
      throw err;
    }
  }
}

// Singleton instance.
export const memoryService = new MemoryService();
